File: src/mod11_gringotts_concurrency/lesson5_countDownLatch/java/Main.java
package mod11_gringotts_concurrency.lesson5_countDownLatch.java;


import java.util.concurrent.CountDownLatch;

public class Main {
    public static void main(String[] args) {

        BankAccount ollivander = new BankAccount("Ollivander's account", 15_000);
        BankAccount hagrid = new BankAccount("Hadrid's account", 500_000);

        CountDownLatch latch=new CountDownLatch(200);

        Thread thread1 = new Thread1(hagrid, ollivander,latch);
        Thread thread2 = new Thread(new RunnableForThread2(hagrid,ollivander,latch));

        thread1.start();
        thread2.start();

//        try {
//            thread1.join();
//            thread2.join();
//        } catch (InterruptedException e) {
//            throw new RuntimeException(e);
//        }


        try {
            latch.await();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }

        System.out.println("Balances after transfers:" + hagrid.getAccountName() + ":" + hagrid.getBalance() + ";" +
                ollivander.getAccountName() + ":" + ollivander.getBalance());
    }
}

File: src/mod11_gringotts_concurrency/lesson5_countDownLatch/kotlin/Main.kt
package mod11_gringotts_concurrency.lesson5_countDownLatch.kotlin

import java.util.concurrent.CountDownLatch
import kotlin.concurrent.thread

fun main() {

    val hagridAccount = BankAccount("Hadrid's account", 500_000)
    val ollivanderAccount = BankAccount("Ollivander's account", 15_000)
    val countDownLatch=CountDownLatch(2)

    val thread1 = thread {
        repeat(100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
        }
        countDownLatch.countDown()
    }
    val thread2 = thread {
        repeat(100) {
            hagridAccount.transferFunds(ollivanderAccount, 10)
        }
        countDownLatch.countDown()
    }
//    thread1.join()
//    thread2.join()

    countDownLatch.await()

    println(
        "Balances after transfers:" + hagridAccount.accountName + ":" + hagridAccount.balance + ";" +
                ollivanderAccount.accountName + ":" + ollivanderAccount.balance
    )


}